packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -997,3 +959,1336 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+        overlayLocation: null,
+        overlayChild: null,
+        child: widget.child,
+      );
+    }
+    return _OverlayPortal(
+      overlayLocation: _location ??= _OverlayEntryLocation(timestamp, widget._resolveLocation(context)._overlayEntryTopChildModel),
+      overlayChild: _DeferredLayout(child: widget.overlayChildBuilder(context)),
+      child: widget.child,
+    );
+  }
+}
+
+// An interface for storing render children of a _RenderTheatre that should be
+// painted together.
+abstract class _RenderTheatreChildModel {
+  // Returns an iterator that traverse the children in the child model in paint
+  // order (from farthest to the user to the closest to the user).
+  //
+  // The returned iterator must be used immediately, before any changes can be
+  // made to the child model. Also the iterator should be safe to use even when
+  // the child model is being mutated. The reason for that is it's allowed to
+  // add/remove/move deferred children to a _RenderTheatre during performLayout,
+  // but the affected children don't have to be laid out in the same
+  // performLayout call.
+  _ChildIterator? _paintOrderIterator();
+
+  // Returns an iterator that traverse the children in the child model in
+  // hit-test order (from closest to the user to the farthest to the user).
+  _ChildIterator? _hitTestOrderIterator();
+
+  void visitChildren(RenderObjectVisitor visitor) {
+    final _ChildIterator? iterator = _paintOrderIterator();
+    if (iterator == null) {
+      return;
+    }
+    RenderBox? child = iterator();
+    while (child != null) {
+      visitor(child);
+      child = iterator();
+    }
+  }
+
+  bool hitTestChildren(BoxHitTestResult result, Offset position) {
+    final _ChildIterator? iterator = _hitTestOrderIterator();
+    if (iterator == null) {
+      return false;
+    }
+    RenderBox? child = iterator();
+    while (child != null) {
+      final StackParentData childParentData = child.parentData! as StackParentData;
+      final bool isHit = result.addWithPaintOffset(
+        offset: childParentData.offset,
+        position: position,
+        hitTest: _toBoxHitTest(child.hitTest)
+      );
+      if (isHit) {
+        return true;
+      }
+      child = iterator();
+    }
+    return false;
+  }
+
+  void paintChildren(PaintingContext context, Offset offset) {
+    final _ChildIterator? iterator = _paintOrderIterator();
+    if (iterator == null) {
+      return;
+    }
+    RenderBox? child = iterator();
+    while(child != null) {
+      final StackParentData parentData = child.parentData! as StackParentData;
+      context.paintChild(child, parentData.offset + offset);
+      child = iterator();
+    }
+  }
+}
+
+// The data structure for managing the stack of theatre children that are sorted
+// by their _showTimestamp.
+//
+// The children with more recent timestamps (i.e. those called `show` recently)
+// will be painted last.
+class _EventOrderChildModel extends _RenderTheatreChildModel {
```

Comment:
Any particular reason for the seperation between _RenderTheatreChildModel and _EventOrderChildModel? It seems it's the only subclass?

Reply:
yeah I'm planning to add more subclasses in the future.